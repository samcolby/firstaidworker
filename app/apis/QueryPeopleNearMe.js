import React from "react";
import PropTypes from "prop-types";

import { Query } from "react-apollo";
import { gql } from "apollo-boost";

const QUERY_PEOPLE_NEAR_ME_DATA = "workers_near_location";
/**
 * GraphQL query to get the list of people nearest the user
 * Used by the QueryPeopleNearMe component
 */
const QUERY_PEOPLE_NEAR_ME = gql`
  query WorkersNearLocation($latitude: String!, $longitude: String!) {
    ${QUERY_PEOPLE_NEAR_ME_DATA}(args: {latitude: $latitude, longitude: $longitude}) {
      id
      name
      job_title
      avatar_uri
      location
    }
  }
`;

const SEARCH_PEOPLE_DATA = "search_workers";
/**
 * GraphQL query to search for active people
 * Used by the QueryPeopleNearMe component
 */
const SEARCH_PEOPLE = gql`
  query($searchquery: String!) {
    ${SEARCH_PEOPLE_DATA}(args: { search: $searchquery }) {
      id
      name
      job_title
      avatar_uri
    }
  }
`;

/**
 * Object containing the different queryTypes that can be used
 */
const QUERY_PEOPLE_NEAR_ME_TYPE = {
  SEARCH: 0,
  QUERY: 1
};

/**
 * Returns the correct configuration for the Apollo Query Component
 *
 * @param {number} queryType
 *    The QUERY_PEOPLE_NEAR_ME_TYPE value to be used to specify the type of query to call
 * @param {string} searchQuery
 *    The string to search for if queryType === QUERY_PEOPLE_NEAR_ME_TYPE.SEARCH
 */
function getQueryConfig(queryType, searchQuery, latitude, longitude) {
  if (queryType === QUERY_PEOPLE_NEAR_ME_TYPE.SEARCH) {
    return {
      gqlQuery: SEARCH_PEOPLE,
      gqlVariables: { searchquery: searchQuery }
    };
  } else {
    return {
      gqlQuery: QUERY_PEOPLE_NEAR_ME,
      gqlVariables: {
        latitude: latitude.toString(),
        longitude: longitude.toString()
      }
    };
  }
}

/**
 * Apollo query component to return a list of people to display in a FlatList
 * This component can execute either a standard query of a search query
 * depending on the queryType value passed in as a prop
 *
 * @param {Object} props - Standard react prop
 * @param {function} children - Standard react prop, doesn't need to be included
 * @param {number} queryType
 *    The QUERY_PEOPLE_NEAR_ME_TYPE value to be used to specify the type of query to call
 * @param {string} searchQuery
 *    The string to search for if queryType === QUERY_PEOPLE_NEAR_ME_TYPE.SEARCH
 */
function QueryPeopleNearMe({
  children,
  queryType,
  searchQuery,
  latitude,
  longitude
}) {
  const config = getQueryConfig(queryType, searchQuery, latitude, longitude);

  return (
    <Query
      fetchPolicy="cache-and-network"
      query={config.gqlQuery}
      variables={config.gqlVariables}
    >
      {props => children(props)}
    </Query>
  );
}

QueryPeopleNearMe.propTypes = {
  children: PropTypes.func,
  queryType: PropTypes.number.isRequired,
  searchQuery: PropTypes.string.isRequired,
  latitude: PropTypes.number.isRequired,
  longitude: PropTypes.number.isRequired
};

export default QueryPeopleNearMe;
export {
  QUERY_PEOPLE_NEAR_ME_TYPE,
  QUERY_PEOPLE_NEAR_ME,
  QUERY_PEOPLE_NEAR_ME_DATA,
  SEARCH_PEOPLE,
  SEARCH_PEOPLE_DATA
};
