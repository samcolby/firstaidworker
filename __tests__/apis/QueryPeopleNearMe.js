import React from "react";
import { Text } from "react-native";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import { MockedProvider } from "react-apollo/test-utils";

import wait from "waait";

import QueryPeopleNearMe, {
  QUERY_PEOPLE_NEAR_ME_TYPE
} from "../../app/apis/QueryPeopleNearMe";

import {
  mocksForQueryPeopleNearMe,
  mocksForQueryPeopleNearMeError,
  mocksForSearchPeople,
  mocksForSearchPeopleError
} from "../mocksdata/MocksQueryPeopleNearMe";

describe("<QueryPeopleNearMe queryType={query}/> component", () => {
  it("Renders correctly", () => {
    const component = shallow(
      <QueryPeopleNearMe
        queryType={QUERY_PEOPLE_NEAR_ME_TYPE.QUERY}
        searchQuery=""
      />
    );

    expect(component.length).toBe(1);
  });

  it("Renders loading correctly", () => {
    const component = renderer.create(
      <MockedProvider mocks={mocksForQueryPeopleNearMe} addTypename={false}>
        <QueryPeopleNearMe
          queryType={QUERY_PEOPLE_NEAR_ME_TYPE.QUERY}
          searchQuery=""
        >
          {({ data, error, fetchMore, loading, networkStatus, refetch }) => {
            if (error) {
              return <Text>error</Text>;
            } else if (loading) {
              return <Text>Loading</Text>;
            } else {
              return <Text>{JSON.stringify(data.profile)}</Text>;
            }
          }}
        </QueryPeopleNearMe>
      </MockedProvider>
    );

    const textComponent = component.root.findByType("Text");
    expect(textComponent.children[0]).toEqual("Loading");

    expect(component).toMatchSnapshot();
  });

  it("Renders data correctly", async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocksForQueryPeopleNearMe} addTypename={false}>
        <QueryPeopleNearMe
          queryType={QUERY_PEOPLE_NEAR_ME_TYPE.QUERY}
          searchQuery=""
        >
          {({ data, error, fetchMore, loading, networkStatus, refetch }) => {
            if (error) {
              return <Text>Error: {JSON.stringify(error)}</Text>;
            } else if (loading) {
              return <Text>loading</Text>;
            } else {
              return <Text>{JSON.stringify(data.profile)}</Text>;
            }
          }}
        </QueryPeopleNearMe>
      </MockedProvider>
    );

    await wait(0);

    const textComponent = component.root.findByType("Text");
    expect(JSON.parse(textComponent.children)[0].name).toEqual(
      "Kaitlin Burnett"
    );

    expect(component).toMatchSnapshot();
  });

  it("Throws errors", async () => {
    const component = renderer.create(
      <MockedProvider
        mocks={mocksForQueryPeopleNearMeError}
        addTypename={false}
      >
        <QueryPeopleNearMe
          queryType={QUERY_PEOPLE_NEAR_ME_TYPE.QUERY}
          searchQuery=""
        >
          {({ data, error, fetchMore, loading, networkStatus, refetch }) => {
            if (error) {
              return <Text>ERROR!</Text>;
            } else if (loading) {
              return <Text>loading</Text>;
            } else {
              return <Text>{JSON.stringify(data.profile)}</Text>;
            }
          }}
        </QueryPeopleNearMe>
      </MockedProvider>
    );

    await wait(0); // wait for response

    const textComponent = component.root.findByType("Text");
    expect(textComponent.children[0]).toContain("ERROR!");
  });
});

describe("<QueryPeopleNearMe queryType={search}/> component", () => {
  it("Renders correctly", () => {
    const component = shallow(
      <QueryPeopleNearMe
        queryType={QUERY_PEOPLE_NEAR_ME_TYPE.SEARCH}
        searchQuery="This is a test string"
      />
    );

    expect(component.length).toBe(1);
  });

  it("Renders loading correctly", () => {
    const component = renderer.create(
      <MockedProvider mocks={mocksForSearchPeople} addTypename={false}>
        <QueryPeopleNearMe
          queryType={QUERY_PEOPLE_NEAR_ME_TYPE.SEARCH}
          searchQuery="DummyQuery"
        >
          {({ data, error, fetchMore, loading, networkStatus, refetch }) => {
            if (error) {
              return <Text>error</Text>;
            } else if (loading) {
              return <Text>Loading</Text>;
            } else {
              return <Text>{JSON.stringify(data.search_profile)}</Text>;
            }
          }}
        </QueryPeopleNearMe>
      </MockedProvider>
    );

    const textComponent = component.root.findByType("Text");
    expect(textComponent.children[0]).toEqual("Loading");

    expect(component).toMatchSnapshot();
  });

  it("Renders data correctly", async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocksForSearchPeople} addTypename={false}>
        <QueryPeopleNearMe
          queryType={QUERY_PEOPLE_NEAR_ME_TYPE.SEARCH}
          searchQuery="DummyQuery"
        >
          {({ data, error, fetchMore, loading, networkStatus, refetch }) => {
            if (error) {
              return <Text>Error: {JSON.stringify(error)}</Text>;
            } else if (loading) {
              return <Text>loading</Text>;
            } else {
              return <Text>{JSON.stringify(data.search_profile)}</Text>;
            }
          }}
        </QueryPeopleNearMe>
      </MockedProvider>
    );

    await wait(0);

    const textComponent = component.root.findByType("Text");
    expect(JSON.parse(textComponent.children)[0].name).toEqual(
      "Maryellen Baxter"
    );

    expect(component).toMatchSnapshot();
  });

  it("Throws errors", async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocksForSearchPeopleError} addTypename={false}>
        <QueryPeopleNearMe
          queryType={QUERY_PEOPLE_NEAR_ME_TYPE.SEARCH}
          searchQuery="DummyQuery"
        >
          {({ data, error, fetchMore, loading, networkStatus, refetch }) => {
            if (error) {
              return <Text>ERROR!</Text>;
            } else if (loading) {
              return <Text>loading</Text>;
            } else {
              return <Text>{JSON.stringify(data.search_profile)}</Text>;
            }
          }}
        </QueryPeopleNearMe>
      </MockedProvider>
    );

    await wait(0);

    const textComponent = component.root.findByType("Text");
    expect(textComponent.children[0]).toContain("ERROR!");
  });
});
