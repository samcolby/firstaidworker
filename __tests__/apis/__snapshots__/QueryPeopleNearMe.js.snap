// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QueryPeopleNearMe queryType={query}/> component Renders data correctly 1`] = `
<Text>
  [{"id":"9f47bfa4-892e-4abd-8a18-0576f8759051","name":"Kaitlin Burnett","job_title":"Marketing","avatar_uri":"https://randomuser.me/api/portraits/women/17.jpg","location":{"coordinates":[-78.5011,-35.033]}},{"id":"0b8a6373-4530-41d7-87a9-435552dfa334","name":"Sargent Bush","job_title":"Sales","avatar_uri":"https://randomuser.me/api/portraits/men/32.jpg","location":{"coordinates":[-78.5011,-35.033]}},{"id":"54f58dfd-3f17-427c-b1d1-74c26a77f700","name":"Key Whitley","job_title":"Michael Jackson Impersonator","avatar_uri":"https://randomuser.me/api/portraits/women/45.jpg","location":{"coordinates":[-78.5011,-35.033]}}]
</Text>
`;

exports[`<QueryPeopleNearMe queryType={query}/> component Renders loading correctly 1`] = `
<Text>
  Loading
</Text>
`;

exports[`<QueryPeopleNearMe queryType={search}/> component Renders data correctly 1`] = `
<Text>
  [{"id":"c4ab9532-198e-4891-8719-8167005e9fa9","name":"Maryellen Baxter","job_title":"Flouncer","avatar_uri":"https://randomuser.me/api/portraits/women/12.jpg"},{"id":"1b773d36-b5c4-4ea9-982a-5ba319dc7bb7","name":"Fry Newton","job_title":"Michael Jackson Impersonator","avatar_uri":"https://randomuser.me/api/portraits/men/65.jpg"},{"id":"228ce656-08ad-46cc-97f8-b4abe71b7b67","name":"Lily Watson","job_title":"Sales","avatar_uri":"https://randomuser.me/api/portraits/women/32.jpg"}]
</Text>
`;

exports[`<QueryPeopleNearMe queryType={search}/> component Renders loading correctly 1`] = `
<Text>
  Loading
</Text>
`;
